src/core/org/apache/jmeter/gui/action/HtmlReportGenerator.java

```diff
@@ -0,0 +1,186 @@
+    public static final String NOT_EMPTY_DIRECTORY = "generate_report_ui.directory_not_empty";
+    public static final String CANNOT_CREATE_DIRECTORY = "generate_report_ui.cannot_create_directory";
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(HtmlReportGenerator.class);
+    private static final long COMMAND_TIMEOUT = JMeterUtils.getPropDefault("generate_report_ui.generation_timeout", 120000L);
+
+    private String csvFilePath;
+    private String userPropertiesFilePath;
+    private String outputDirectoryPath;
+
+    public HtmlReportGenerator(String csvFilePath, String userPropertiesFilePath, String outputDirectoryPath) {
+        this.csvFilePath = csvFilePath;
+        this.userPropertiesFilePath = userPropertiesFilePath;
+        if (outputDirectoryPath == null) {
+            this.outputDirectoryPath = JMeterUtils.getJMeterBinDir() + "/report-output/";
+        } else {
+            this.outputDirectoryPath = outputDirectoryPath;
+        }
+    }
+
+    /*
+     * Prepare and Run the HTML report generation command
+     */
+    public List<String> run() {
+        List<String> errorMessageList = new ArrayList<>();
+        errorMessageList.addAll(checkArguments());
+        if (!errorMessageList.isEmpty()) {
+            return errorMessageList;
+        }
+        
+        ByteArrayOutputStream commandExecutionOutput = new ByteArrayOutputStream();
+        int resultCode = -1;
+        List<String> generationCommand = createGenerationCommand();
+        try {
+            SystemCommand sc = new SystemCommand(new File(JMeterUtils.getJMeterBinDir()), COMMAND_TIMEOUT, 100, null, null,
+                    commandExecutionOutput, null);
+            LOGGER.debug("Running report generation");
+            resultCode = sc.run(generationCommand);
+            if (resultCode != 0) {
+                errorMessageList.add(commandExecutionOutput.toString());
+                LOGGER.info("The HTML report generation failed and returned: {}", commandExecutionOutput);
+                return errorMessageList;
+            }
+        } catch (InterruptedException | IOException e) {
+            errorMessageList.add(commandExecutionOutput.toString());
+            if (LOGGER.isErrorEnabled()) {
+                LOGGER.error("Error during HTML report generation: {}", e.getMessage(), e);
+            }
+        }
+        LOGGER.debug("SystemCommand ran: {}  returned: {}", generationCommand, resultCode);
+        return errorMessageList;
+    }
+
+    /**
+     * create the command for html report generation with all the directories /
+     * file
+     * 
+     * @return the list of arguments for SystemCommand execution
+     */
+    private List<String> createGenerationCommand() {
+        String jmeterBinDir = JMeterUtils.getJMeterBinDir();
+        List<String> arguments = new ArrayList<>();
+        String java = System.getProperty("java.home") + "/bin/java";
```

Comment:
What's the reason to startup external processes?

Can't current JVM be used to generate the required reports?

Reply:
Some properties (save service) are loaded and stored as constants.
So if user used different set of properties for generating the report, they would not be overridable.

Reply:
Well, debugging through this shell command is hard.